top: Common network hostname/country configuration.

jimmo opened the pull request:
The overall goal is to allow configuration of the hostname on Pico W, but also unify the way networking works across ports. Fixes #8906, #10397

This PR adds `network.hostname()` and `network.country()` to globally set the device hostname (for all interfaces) and country (for radio compliance). For interfaces that previously supported `nic.config(hostname)` this will continue to work, as will `pyb.country()` (and `rp2.country()`).

Additionally configure the default hostname on a per-port and per-board basis. We may want to remove this change if it's considered a breaking change to change the default hostname?

It also updates the stm32 port to use the new cyw43-driver (in the lib/cyw43-driver submodule) and removes the old version from drivers. This relies on https://github.com/georgerobotics/cyw43-driver/pull/58.

Next steps (separate PR):
 - Investigate configuring mDNS more explicitly. See #9058 and #9086 for more history.
 - Implement hostname for other NICs (Nina, Wiznet) and country for all other non-cyw43 wifi (Nina, ESP8266, ESP32).
 - Unify other networking config, in particular the auth and mode constants, and make it possible to write the same code on all ports to manage the WiFi interface. Eventually I'd like to remove the modnetwork customisations used by ESP8266/ESP32 and have them work like the other ports.

jimmo committed:
extmod/modnetwork: Add network.hostname() and network.country().

This provides a standard interface to setting the global networking config
for all interfaces and interface types.

For ports that already use either a static hostname (mimxrt, rp2) they will
now use the configured value. The default is configured by the port
(or optionally the board).

For interfaces that previously supported .config(hostname), this is still
supported but now implemented using the global network.hostname.

Similarly, pyb.country and rp2.country are now deprecated, but the methods
still exist (and forward to network.hostname).

Because ESP32/ESP8266 do not use extmod/modnetwork.c they are not affected
by this commit.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modnetwork: Allow more extensive port-specific customisation.

This allows for a port (e.g. esp8266/esp32) to use extmod/modnetwork.c
and provide the globals dict, rather than just a list of interfaces.

When this is used, the default implementation of `network.route` and the
NIC list is not enabled.

Also splits out the LWIP-specific helpers from modnetwork.c into
network_lwip.c.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp32: Use extmod/modnetwork.c instead of port-specific version.

Rather than duplicating the implementation of `network`, this allows ESP32
to use the shared one in extmod.  In particular this gains access to
network.hostname and network.country.

Set default hostnames for various ESP32 boards.

Other than adding these two methods and the change to the default hostname,
there is no other user-visible change.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp8266: Use extmod/modnetwork.c instead of port-specific version.

Rather than duplicating the implementation of `network`, this allows
ESP8266 to use the shared one in extmod.  In particular this gains access
to network.hostname and network.country.

Other than adding these two methods, there is no other user-visible change.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32: Update to use the open-source lib version of cyw43-driver.

This removes the previous WiFi driver from drivers/cyw43 (but leaves behind
the BT driver), and makes the stm32 port (i.e. PYBD and Portenta) use the
new "lib/cyw43-driver" open-source driver already in use by the rp2 port.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/network: Update docs for network.country, network.hostname.

Also marks wlan.config(hostname) as deprecated.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
This removes `pyb.country` for boards that don't have networking... Is that going to be a problem?

dpgeorge commented:
> This removes `pyb.country` for boards that don't have networking... Is that going to be a problem?

It might be... the default `boot.py` has a `pyb.country()` entry regardless of networking support or not.  So a lot of boards are still going to have that line in their `boot.py`.  The idea was that even on a PYBv1.1 you might want to configure the country.

jimmo committed:
stm32: Add a default implementation of pyb.country.

This is for boards without networking support so that the default boot.py
continues to work.

Also update boot.py to use network.country and network.hostname instead.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> It might be... the default `boot.py` has a `pyb.country()` entry regardless of networking support or not. So a lot of boards are still going to have that line in their `boot.py`. The idea was that even on a PYBv1.1 you might want to configure the country.

OK. I've added a "no-op" version of pyb.country for boards that don't have networking support so that existing boards with the default boot.py don't break. I've also updated the default boot.py to use network.country and network.hostname instead.

It seems unnecessary to add the whole network module for non-networking boards just to have a .country method that does nothing.

jimmo commented:
Updated the submodule, including removing the cyw43_resource.o handling in favour of the firmware header files. Note that this breaks Pico W until the next Pico SDK update (at which point I will update this PR).

dpgeorge commented:
> Note that this breaks Pico W until the next Pico SDK update

See #10764 which does the update to pico-sdk.

dpgeorge commented:
Please rebase on latest master to get the new pico-sdk (and cyw43-driver).

jimmo commented:
> Please rebase on latest master to get the new pico-sdk (and cyw43-driver).

Done.

dpgeorge committed:
stm32/boards: Update linker scripts now that big_const is gone.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
This now passes the CI.

I tested BLE and WiFi on PYBD-SF2W and it looks good, but I want to do more tests.

dpgeorge committed:
extmod/network_cyw43: Use CYW43_CHANNEL_NONE for default arg in connect.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented on extmod/modnetwork.c:
```diff
@@ -89,9 +89,52 @@ STATIC mp_obj_t network_route(void) {
 STATIC MP_DEFINE_CONST_FUN_OBJ_0(network_route_obj, network_route);
 
+
+char mod_network_country_code[2] = "XX";
```
this variable definition should go at the top of the file, just before the start of function definitons

dpgeorge commented on extmod/modnetwork.c:
```diff
@@ -89,9 +89,52 @@ STATIC mp_obj_t network_route(void) {
+#error "MICROPY_PY_NETWORK_DEFAULT_HOSTNAME must be set in mpconfigport.h or mpconfigboard.h"
+#endif
+
+char mod_network_hostname[MICROPY_PY_NETWORK_MAX_HOSTNAME_LEN] = MICROPY_PY_NETWORK_DEFAULT_HOSTNAME;
```
move to top of file

dpgeorge replied:
also, as above, this could be named `MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN`

dpgeorge commented on extmod/modnetwork.c:
```diff
@@ -89,9 +89,52 @@ STATIC mp_obj_t network_route(void) {
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_network_country_obj, 0, 1, network_country);
+
+#ifndef MICROPY_PY_NETWORK_DEFAULT_HOSTNAME
+#error "MICROPY_PY_NETWORK_DEFAULT_HOSTNAME must be set in mpconfigport.h or mpconfigboard.h"
```
how about naming this `MICROPY_PY_NETWORK_HOSTNAME_DEFAULT`?  the point being that it's a configuration of `network.hostname`

dpgeorge commented on extmod/modnetwork.h:
```diff
@@ -106,4 +106,12 @@ void mod_network_deinit(void);
+#define MICROPY_PY_NETWORK_MAX_HOSTNAME_LEN (16)
+#endif
+
+extern char mod_network_hostname[MICROPY_PY_NETWORK_MAX_HOSTNAME_LEN];
```
same as in the .c file, these should go before function definitions

dpgeorge commented on extmod/network_cyw43.c:
```diff
@@ -119,14 +120,21 @@ STATIC mp_obj_t network_cyw43_deinit(mp_obj_t self_in) {
     } else {
         #if MICROPY_PY_NETWORK_CYW43_USE_LIB_DRIVER
-        cyw43_wifi_set_up(self->cyw, self->itf, mp_obj_is_true(args[1]), MICROPY_CYW43_COUNTRY);
+        uint32_t country = CYW43_COUNTRY(mod_network_country_code[0] || 'X', mod_network_country_code[1] || 'X', 0);
```
is it necessary to do the `|| 'X'`?  isn't it already checked that the value set by the user is a string of length 2?

dpgeorge commented on ports/rp2/boards/W5100S_EVB_PICO/mpconfigboard.h:
```diff
@@ -1,8 +1,9 @@
-#define MICROPY_HW_BOARD_NAME             "W5100S-EVB-Pico"
-#define MICROPY_HW_FLASH_STORAGE_BYTES    (1408 * 1024)
+#define MICROPY_HW_BOARD_NAME               "W5100S-EVB-Pico"
+#define MICROPY_PY_NETWORK_DEFAULT_HOSTNAME "W5100S-EVB"
```
I usually try not to mix HW and PY config options.  Maybe this can go down just below `MICROPY_PY_NETWORK`?

jimmo commented:
Addressed all comments.

dpgeorge commented:
Thank you, this looks good now.  I rebased and updated some of the commit messages (minor changes) and force pushed to this PR branch.

robert-hh commented:
Thanks @jimmo. That was quite a piece of work, and needed.

DataMinded commented:
can we add a few DHCP options in the made discover and offer requests ?
If I capture an raspberry pi 3 with hostname set  DHCP discover and request packages ,I can see it includes a few more parameters, and options in the 55- Parameter Request list .
The RPI is accessable by other devices within the same network by hostname.local  . this is not true for the current micropython dhcp handshake ,  can we add these options to it ??   Unsure where to look to try it myself

Check images below , left is RPI , right is Pico W
![image](https://github.com/micropython/micropython/assets/77164467/11360fbe-79b8-4b33-a908-e407e14c9ff2)


![image](https://github.com/micropython/micropython/assets/77164467/89865f76-7f11-4347-87b9-b1b506fc1a7e)



![image](https://github.com/micropython/micropython/assets/77164467/f9878fdf-3385-4690-b847-061002557413)
I would assume these changes would make the pico W accessable by hostname in local networks
